src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+        self.hide_percentage = hide_percentage
+        self.hide_eta = hide_eta
+
+        self._percentage = None
+        self.total = total
+
+    def compose(self) -> ComposeResult:
+        with Horizontal():
+            if not self.hide_bar:
+                bar = Bar()
+                # Notify the bar when the percentage is updated.
+                self.watch(
+                    self, "_percentage", partial(setattr, bar, "_completion_percentage")
+                )
+                yield bar
+            if not self.hide_percentage:
+                percentage_status = PercentageStatus()
+                # Notify the percentage status label when the percentage is updated.
+                self.watch(
+                    self,
+                    "_percentage",
+                    partial(setattr, percentage_status, "_completion_percentage"),
+                )
+                yield percentage_status
+            if not self.hide_eta:
+                eta_status = ETAStatus()
+                # Notify the ETA status label when the percentage is updated.
+                self.watch(
+                    self,
+                    "_percentage",
+                    partial(setattr, eta_status, "_completion_percentage"),
+                )
+                yield eta_status
+
+    def validate_progress(self, progress: float) -> float:
+        """Prevent the progress to go over the total number of steps."""
+        if self.total is not None:
+            return min(progress, self.total)
+        return progress
+
+    def watch_total(self, total: float | None) -> None:
+        """Post the message [`ProgressBar.Started`][textual.widgets.ProgressBar.Started]."""
+        if total is None:
+            return
+        if not self._started:
+            self.post_message(ProgressBar.Started(self, total))
+            self._started = True
+
+    def compute__percentage(self) -> float | None:
+        """Keep the percentage of progress updated automatically.
+
+        This compute method will also post the message
+        [`ProgressBar.Completed`][textual.widgets.ProgressBar.Completed]
+        when the percentage reaches 1.
+        """
+        if self.total is not None:
+            percentage = self.progress / self.total
+            if percentage >= 1:
+                self.post_message(ProgressBar.Completed(self))
+            return percentage
+        return None
+
+    @property
+    def percentage(self) -> float | None:
+        """The progress percentage, or `None` if not available.
+
+        The percentage is a value between 0 and 1 and the returned value is only
+        `None` if the total progress of the bar hasn't been set yet.
+        In other words, after the progress bar emits the message
+        [`ProgressBar.Started`][textual.widgets.ProgressBar.Started],
+        the value of `percentage` is always not `None`.
+
+        Example:
+            ```py
+            pb = ProgressBar()
```

Comment:
In the other examples a variable called `progress_bar` tends to be used, presumably avoiding the "no abbreviations" rule; perhaps do the same here too?